src/copc.rs

```diff
@@ -0,0 +1,507 @@
+/// node. The size and file offset of each data chunk is provided in the hierarchy
+/// entries, allowing the chunks to be directly read for decoding.
+#[derive(Debug)]
+pub struct CopcHierarchyVlr {
+    root: Page,
+    sub_pages: HashMap<VoxelKey, Page>,
+}
+impl CopcHierarchyVlr {
+    /// The record id of the LasZip VLR header.
+    pub const RECORD_ID: u16 = 1000;
+
+    /// Writes the Vlr data to the source.
+    ///
+    /// This **only** writes the *payload data* the
+    /// vlr header should be written before-hand.
+    pub fn write_to<W: Write>(&self, dst: &mut W) -> Result<()> {
+        self.root.write_to(dst)?;
+        self.sub_pages
+            .iter()
+            .try_for_each(|(_, page)| page.write_to(dst))
+    }
+    /// Reads the CopcHierarchyVlr from the Vlr payload with specifications from copc_info
+    pub fn read_from_with(vlr: &Vlr, copc_info: &CopcInfoVlr) -> Result<CopcHierarchyVlr> {
+        let root = Page::read_from(vlr.data[0..copc_info.root_hier_size as usize].as_ref())?;
+        let sub_pages = root
+            .entries
+            .iter()
+            .filter(|entry| entry.is_referencing_page())
+            .map(|entry| {
+                let start = (entry.offset - copc_info.root_hier_offset) as usize;
+                let end = start + entry.byte_size as usize;
+                Page::read_from(vlr.data[start..end].as_ref()).map(|p| (entry.key, p))
+            })
+            .collect::<Result<HashMap<VoxelKey, Page>>>()?;
+        Ok(CopcHierarchyVlr { root, sub_pages })
+    }
+    /// iterates over all entries merging all referenced pages into root
+    pub fn iter_entrys(&self) -> impl Iterator<Item = Entry> {
+        self.root.entries.iter().flat_map(|entry| {
+            if entry.is_referencing_page() {
+                if let Some(page) = self.sub_pages.get(&entry.key) {
+                    page.entries.clone()
+                } else {
+                    // this entry is corrupt or the page is missing
+                    vec![Entry {
+                        key: entry.key,
+                        offset: entry.offset,
+                        byte_size: 0,
+                        point_count: 0,
+                    }]
+                }
+            } else {
+                vec![*entry]
+            }
+        })
+    }
+}
+impl Vlr {
+    /// Returns true if this [Vlr] is the Copc info Vlr.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #[cfg(feature = "copc")]
+    /// {
+    /// use las::{copc, Vlr};
+    ///
+    /// let mut vlr = Vlr::default();
+    /// assert!(!copc::is_copcinfo_vlr(&vlr));
+    /// vlr.user_id = "copc".to_string();
+    /// vlr.record_id = 1;
+    /// assert!(copc::is_copcinfo_vlr(&vlr));
+    /// }
+    /// ```
+    pub fn is_copc_info(&self) -> bool {
+        self.user_id == USER_ID && self.record_id == CopcInfoVlr::RECORD_ID
+    }
+}
+
+impl Header {
+    /// doc
+    pub fn copc_info_vlr(&self) -> Result<CopcInfoVlr> {
```

Comment:
I _think_ I'm more inclined to return an option, and let the caller decide how bad that is:

```suggestion
    pub fn copc_info_vlr(&self) -> Option<CopcInfoVlr> {
```